### ISSUE TITLE
clang ias: build failure for arch/arm/probes/kprobes/test-arm.c

### ISSUE THREAD

**Issue by arndb** [ClangBuiltLinux/linux#1271]
This is the failure I get for this file when CONFIG_ARM_KPROBES_TEST is enabled

/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:131:2: error: unexpected token in '.ascii' directive
        DATA_PROCESSING_DNM("and",0xf00f00ff)
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:81:2: note: expanded from macro 'DATA_PROCESSING_DNM'
        _DATA_PROCESSING_DNM(op,"",val)         \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:58:2: note: expanded from macro '_DATA_PROCESSING_DNM'
        TEST_RR(  op "eq" s "   r0,  r",1, VAL1,", r",2, val, "")               \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:239:2: note: expanded from macro 'TEST_RR'
        TESTCASE_START(code1 #reg1 code2 #reg2 code3)           \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:113:14: note: expanded from macro 'TESTCASE_START'
        "10:                                            \n\t"   \
                                                          ^
<inline asm>:4:15: note: instantiated into assembly here
        .ascii "and" "eq" "" "  r0,  r" "1" ", r" "2" ""                                
                     ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:131:2: error: unexpected token in '.ascii' directive
        DATA_PROCESSING_DNM("and",0xf00f00ff)
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:81:2: note: expanded from macro 'DATA_PROCESSING_DNM'
        _DATA_PROCESSING_DNM(op,"",val)         \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:59:2: note: expanded from macro '_DATA_PROCESSING_DNM'
        TEST_RR(  op "ne" s "   r1,  r",1, VAL1,", r",2, val, ", lsl #3")       \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:239:2: note: expanded from macro 'TEST_RR'
        TESTCASE_START(code1 #reg1 code2 #reg2 code3)           \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:113:14: note: expanded from macro 'TESTCASE_START'
        "10:                                            \n\t"   \
                                                          ^
<inline asm>:4:15: note: instantiated into assembly here
        .ascii "and" "ne" "" "  r1,  r" "1" ", r" "2" ", lsl #3"                                
                     ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:131:2: error: unexpected token in '.ascii' directive
        DATA_PROCESSING_DNM("and",0xf00f00ff)
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:81:2: note: expanded from macro 'DATA_PROCESSING_DNM'
        _DATA_PROCESSING_DNM(op,"",val)         \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:60:2: note: expanded from macro '_DATA_PROCESSING_DNM'
        TEST_RR(  op "cs" s "   r2,  r",3, VAL1,", r",2, val, ", lsr #4")       \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:239:2: note: expanded from macro 'TEST_RR'
        TESTCASE_START(code1 #reg1 code2 #reg2 code3)           \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:113:14: note: expanded from macro 'TESTCASE_START'
        "10:                                            \n\t"   \
                                                          ^
<inline asm>:4:15: note: instantiated into assembly here
        .ascii "and" "cs" "" "  r2,  r" "3" ", r" "2" ", lsr #4"                                
                     ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:131:2: error: unexpected token in '.ascii' directive
        DATA_PROCESSING_DNM("and",0xf00f00ff)
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:81:2: note: expanded from macro 'DATA_PROCESSING_DNM'
        _DATA_PROCESSING_DNM(op,"",val)         \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-arm.c:61:2: note: expanded from macro '_DATA_PROCESSING_DNM'
        TEST_RR(  op "cc" s "   r3,  r",3, VAL1,", r",2, val, ", asr #5")       \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:239:2: note: expanded from macro 'TEST_RR'
        TESTCASE_START(code1 #reg1 code2 #reg2 code3)           \
        ^
/git/arm-soc/arch/arm/probes/kprobes/test-core.h:113:14: note: expanded from macro 'TESTCASE_START'
        "10:                                            \n\t"   \
                                                          ^
<inline asm>:4:15: note: instantiated into assembly here
        .ascii "and" "cc" "" "  r3,  r" "3" ", r" "2" ", asr #5"                                
                     ^

**Comment by nickdesaulniers** [comment]
cc @jcai19

**Comment by nickdesaulniers** [comment]
is this with clang-11? I get a different set of failures in this TU with clang-12 for allmodconfig.

```
  CC [M]  arch/arm/probes/kprobes/test-arm.o
arch/arm/probes/kprobes/test-arm.c:131:2: error: instruction 'andeq' can not set flags, but 's' suffix specified
        DATA_PROCESSING_DNM("and",0xf00f00ff)
        ^
arch/arm/probes/kprobes/test-arm.c:82:2: note: expanded from macro 'DATA_PROCESSING_DNM'
        _DATA_PROCESSING_DNM(op,"s",val)
        ^
arch/arm/probes/kprobes/test-arm.c:58:2: note: expanded from macro '_DATA_PROCESSING_DNM'
        TEST_RR(  op "eq" s "   r0,  r",1, VAL1,", r",2, val, "")               \
        ^
arch/arm/probes/kprobes/test-core.h:243:2: note: expanded from macro 'TEST_RR'
        TEST_INSTRUCTION(code1 #reg1 code2 #reg2 code3)         \
        ^
arch/arm/probes/kprobes/test-core.h:155:17: note: expanded from macro 'TEST_INSTRUCTION'
        "50:    nop                                     \n\t"   \
                                                          ^
<inline asm>:24:5: note: instantiated into assembly here
        1:      andeqs  r0,  r1, r2                             
                ^
```

**Comment by nickdesaulniers** [comment]
For my case, `-no-integrated-as` works. The disassembly produces `andseq  r0, r1, r2` (from llvm-objdump and GNU objdump).  Looks like gas accepts `andseq` but LLVM_IAS does not.  This pattern happens a lot in this source file.

**Comment by nickdesaulniers** [comment]
I wonder if this is unified syntax issue? I'm re-writing it...slowly...

**Comment by nickdesaulniers** [comment]
[wip.patch.txt](https://github.com/ClangBuiltLinux/linux/files/5884421/wip.patch.txt)

**Comment by nickdesaulniers** [comment]
[wip.patch.txt](https://github.com/ClangBuiltLinux/linux/files/5884513/wip.patch.txt)

**Comment by nickdesaulniers** [comment]
damn, I have this finished, but there is one bug in LLVM that I have a fix for too...

**Comment by nickdesaulniers** [comment]
llvm patch:
```diff
diff --git a/llvm/lib/Target/ARM/ARMInstrInfo.td b/llvm/lib/Target/ARM/ARMInstrInfo.td                                                                                          
index 8dcb319923ae..9792586a53a5 100644
--- a/llvm/lib/Target/ARM/ARMInstrInfo.td
+++ b/llvm/lib/Target/ARM/ARMInstrInfo.td
@@ -3579,10 +3579,10 @@ def MOVr_TC : AsI1<0b1101, (outs tcGPR:$Rd), (ins tcGPR:$Rm), DPFrm,
   let Inst{15-12} = Rd;
 }
 
-def MOVsr : AsI1<0b1101, (outs GPRnopc:$Rd), (ins shift_so_reg_reg:$src),
+def MOVsr : AsI1<0b1101, (outs GPR:$Rd), (ins shift_so_reg_reg:$src),
                 DPSoRegRegFrm, IIC_iMOVsr,
                 "mov", "\t$Rd, $src",
-                [(set GPRnopc:$Rd, shift_so_reg_reg:$src)]>, UnaryDP,
+                [(set GPR:$Rd, shift_so_reg_reg:$src)]>, UnaryDP,
                 Sched<[WriteALU]> {
   bits<4> Rd;
   bits<12> src;
@@ -6327,16 +6327,16 @@ def RRXi : ARMAsmPseudo<"rrx${s}${p} $Rd, $Rm",
                         (ins GPR:$Rd, GPR:$Rm, pred:$p, cc_out:$s)>;
 let TwoOperandAliasConstraint = "$Rn = $Rd" in {
 def ASRr : ARMAsmPseudo<"asr${s}${p} $Rd, $Rn, $Rm",
-                        (ins GPRnopc:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
+                        (ins GPR:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
                              cc_out:$s)>;
 def LSRr : ARMAsmPseudo<"lsr${s}${p} $Rd, $Rn, $Rm",
-                        (ins GPRnopc:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
+                        (ins GPR:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
                              cc_out:$s)>;
 def LSLr : ARMAsmPseudo<"lsl${s}${p} $Rd, $Rn, $Rm",
-                        (ins GPRnopc:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
+                        (ins GPR:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
                              cc_out:$s)>;
 def RORr : ARMAsmPseudo<"ror${s}${p} $Rd, $Rn, $Rm",
-                        (ins GPRnopc:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
+                        (ins GPR:$Rd, GPRnopc:$Rn, GPRnopc:$Rm, pred:$p,
                              cc_out:$s)>;
 }
```
kernel patch: 
[wip.patch.txt](https://github.com/ClangBuiltLinux/linux/files/5884711/wip.patch.txt)

**Comment by nickdesaulniers** [comment]
https://reviews.llvm.org/D95586

**Comment by nickdesaulniers** [comment]
https://lore.kernel.org/lkml/20210128044356.38419-1-ndesaulniers@google.com/T/#u

**Comment by jcai19** [comment]
@arndb I tried with stock LLVM and it was able to assemble ```.ascii "and" "cs" "" " r2, r" "3" ", r" "2" ", lsr #458```. The support to space in .ascii was first added to IAS in https://github.com/llvm/llvm-project/commit/e0963ae274be, but got temporarily reverted at https://github.com/llvm/llvm-project/commit/82c4153e66fa for a test failure (which turned out to be a false-positive). It was then relanded again in https://github.com/llvm/llvm-project/commit/9dfeec853008109b1cbe926c22675c96226040d9. Is it possible that the Clang version you used fell into this range?

**Comment by arndb** [comment]
On Thu, Jan 28, 2021 at 10:10 AM Jian Cai <notifications@github.com> wrote:

> @arndb <https://github.com/arndb> I tried with stock LLVM and it was able
> to assemble .ascii "and" "cs" "" " r2, r" "3" ", r" "2" ", lsr #458. The
> support to space in .ascii was first added to IAS in llvm/llvm-project@
> e0963ae <https://github.com/llvm/llvm-project/commit/e0963ae274be>, but
> got temporarily reverted at llvm/llvm-project@82c4153
> <https://github.com/llvm/llvm-project/commit/82c4153e66fa> for a test
> failure (which turned out to be a false-positive). It was then relanded
> again in llvm/llvm-project@9dfeec8
> <https://github.com/llvm/llvm-project/commit/9dfeec853008109b1cbe926c22675c96226040d9>.
> Is it possible that the Clang version you used fell into this range?
>

I was initially building with clang-11  by accident, retrying with Ubuntu
clang version
12.0.0-++20210126052625+879c12d95a27-1~exp1~20210126163340.296 now.

**Comment by nickdesaulniers** [comment]
@arndb noted that arch/arm/probes/kprobes/test-thumb.c also has a similar issue with CONFIG_THUMB2_KERNEL.

**Comment by nickdesaulniers** [comment]
v5: https://lore.kernel.org/linux-arm-kernel/20210211025149.3544593-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
submitted:
- https://www.arm.linux.org.uk/developer/patches/viewpatch.php?id=9061/1
- https://www.arm.linux.org.uk/developer/patches/viewpatch.php?id=9062/1

**Comment by nickdesaulniers** [comment]
fixed: e9d54be9ad5ee2eed3056d8901ac4b3b115d95a5
fixed: 7c182ebab9f338d07571c7f86d4d64c385ad1b9c
v5.13-rc1
